common/config/src/properties.rs

```diff
@@ -0,0 +1,233 @@
+use crate::argv::env;
+use crate::error::ConfigError;
+use crate::raw::{Config, Rules};
+use crate::{argv, get_hostname};
+use http::types::params::{Params, Tags};
+use java_properties::PropertiesIter;
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::BufReader;
+use std::path::PathBuf;
+use std::str::FromStr;
+
+macro_rules! from_env_name {
+    ($key: ident) => {
+        static $key: Key = Key::FromEnv(env::$key);
+    };
+}
+
+// Names from the legacy agent
+static INGESTION_KEY: Key = Key::Name("key");
+static LOG_DIRS: Key = Key::Name("logdir");
+static TAGS: Key = Key::Name("tags");
+static OS_HOSTNAME: Key = Key::Name("hostname");
+static EXCLUSION_RULES: Key = Key::Name("exclude");
+static EXCLUSION_REGEX_RULES: Key = Key::Name("exclude_regex");
+
+// New keys: reuse the same name from env vars, removing the LOGDNA_ prefix and
+// setting it to lowercase
+from_env_name!(ENDPOINT);
+from_env_name!(HOST);
+from_env_name!(USE_SSL);
+from_env_name!(USE_COMPRESSION);
+from_env_name!(GZIP_LEVEL);
+from_env_name!(INCLUSION_RULES);
+from_env_name!(INCLUSION_REGEX_RULES);
+from_env_name!(IP);
+from_env_name!(MAC);
+from_env_name!(JOURNALD_PATHS);
+from_env_name!(LOOKBACK);
+from_env_name!(DB_PATH);
+from_env_name!(METRICS_PORT);
+from_env_name!(USE_K8S_LOG_ENRICHMENT);
+from_env_name!(LOG_K8S_EVENTS);
+from_env_name!(LINE_EXCLUSION);
+from_env_name!(LINE_INCLUSION);
+from_env_name!(REDACT);
+
+enum Key {
+    FromEnv(&'static str),
```

Comment:
Given these are both wrapping up &'static str would it make sense to do

```
struct Key(String);

impl FromStr for Key {
    fn parse(...: &'str) -> Self {
        ...
    }
}

```
That would let us parse once on creation of the key and also let us drop the static bounds

Reply:
Although, reading it, it would require the from_env_name to be lazy statics wouldn't it. We might be able to use a const fn to build the new keys...

Reply:
Nope! not yet, and the various macro libraries don't seem to make this possible either. Comments retracted as not worthwhile!

Reply:
yeah, I give const fn a try and couldn't do it :)